packages/core/saved-objects/core-saved-objects-api-server-internal/src/lib/repository.ts

```
@@ -762,6 +776,289 @@ export class SavedObjectsRepository implements ISavedObjectsRepository {
+            ),
+            _index: this.getIndexForType(expectedResult.value.type),
+            ...getExpectedVersionProperties(undefined),
+          },
+        });
+      }
+    });
+
+    const bulkDeleteResponse = bulkDeleteParams.length
+      ? await this.client.bulk({
+          refresh,
+          body: bulkDeleteParams,
+          require_alias: true,
+        })
+      : undefined;
+
+    // extracted to ensure consistency in the error results returned
+    let errorResult: BulkDeleteItemErrorResult;
+    const objectsToDeleteAliasesFor: ObjectToDeleteAliasesFor[] = [];
+
+    const savedObjects = expectedBulkDeleteMultiNamespaceDocsResults.map((expectedResult) => {
+      if (isLeft(expectedResult)) {
+        return { ...expectedResult.value, success: false };
+      }
+      const {
+        type,
+        id,
+        namespaces,
+        esRequestIndex: esBulkDeleteRequestIndex,
+      } = expectedResult.value;
+      // we assume this wouldn't happen but is needed to ensure type consistency
+      if (bulkDeleteResponse === undefined) {
+        throw new Error(
+          `Unexpected error in bulkDelete saved objects: bulkDeleteResponse is undefined`
+        );
+      }
+      const rawResponse = Object.values(
+        bulkDeleteResponse.items[esBulkDeleteRequestIndex]
+      )[0] as NewBulkItemResponse;
+
+      const error = getBulkOperationError(type, id, rawResponse);
+      if (error) {
+        errorResult = { success: false, type, id, error };
+        return errorResult;
+      }
+      if (rawResponse.result === 'not_found') {
+        errorResult = {
+          success: false,
+          type,
+          id,
+          error: errorContent(SavedObjectsErrorHelpers.createGenericNotFoundError(type, id)),
+        };
+        return errorResult;
+      }
+
+      if (rawResponse.result === 'deleted') {
+        // `namespaces` should only exist in the expectedResult.value if the type is multi-namespace.
+        if (namespaces) {
```

NIT: this condition should always be true here, right? I'm fine keeping, but then we should throw in a else block as we did for `if (bulkDeleteResponse === undefined) {`

---

> this condition should always be true here, right?
Not necessarily. Only multi-namespace documents have a `namespaces` array because we're only using `mget` for those namespace types. Single namespace and agnostic types aren't fetched from es in the `mget` call.